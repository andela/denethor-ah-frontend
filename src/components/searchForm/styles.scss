form {
  > .flex {
    justify-content: space-between;
  }
}

.css-10nd86i,
.css-1sontr1  {
  width: 10rem;
}

.css-wqgs6e,
.css-v73v8k {
  background: white !important;
  color: black !important;
}

.css-162g8z5,
input:disabled {
  cursor: not-allowed;
}