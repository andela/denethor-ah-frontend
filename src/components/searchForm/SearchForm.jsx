import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import Select from 'react-select';

import getFilteredArticles from '../../redux/actions/filters';
import { getTags } from '../../redux/actions/tags';
import getAuthors from '../../redux/actions/authors';
import './styles.scss'

export class SearchForm extends Component { 
  state = {
    searchString: '',
    selectedTag: '',
    selectedAuthor: '',
    authorOptions: [],
    tagOptions: [],
    disableButton: false,
    toastOptions: {
      hideProgressBar: true,
      closeButton: false,
      toastId: 'searchToast',
      className: 'toast-custom-style'
    }
  }

  async componentDidUpdate(prevProps) {
    const { props: { isLoggedIn, handleGetAuthors, handleGetTags } } = this;

    if (isLoggedIn !== prevProps.isLoggedIn) {
      
      await handleGetAuthors();
      await handleGetTags()

      const authorOptions = this.props.authors.map(({ username }) => ({ value: username, label: username }));
      const tagOptions = this.props.tags.map(({ id, tagText }) => ({ value: id, label: tagText }));

      this.setState({ authorOptions });
      this.setState({ tagOptions });
    }
  }

  onSearchChange = ({ target: { value } }) => this.setState({ searchString: value });
  onAuthorChange = (value) => this.setState({ selectedAuthor: value });
  onTagChange = (value) => this.setState({ selectedTag: value });

  onSubmit = async (e) => {
    e.preventDefault();
    const { 
      state: { searchString, selectedTag, selectedAuthor, toastOptions },
      props: { handleFilter, history } 
    } = this;

    if (!searchString && !selectedTag && !selectedAuthor) {
      return toast.error('No filter parameter was entered', toastOptions);
    }

    this.setState({ disableButton: true });
    await handleFilter(searchString, selectedTag.value, selectedAuthor.value);
    this.setState({ disableButton: false });
    history.push('/filter');
  }

  render() {
    const { 
      props: { isLoggedIn, location: { pathname }, profile: { username = '' } =  {} }, 
      state: { searchString, selectedAuthor, selectedTag, authorOptions, tagOptions, disableButton },
      onSearchChange, onAuthorChange, onTagChange, onSubmit 
    } = this;

    return (
      <form onSubmit={onSubmit}>
        <div className={`search-bar ${!isLoggedIn ? 'w-100' : ''}`}>
          <input
            type='text'
            placeholder={
              pathname.includes("/dashboard") ?
              'You can only filter dashboard publications by Tags' :
              'Search for articles...'
            }
            value={searchString}
            onChange={onSearchChange}
            disabled={pathname.includes("/dashboard")}
          />
        </div>
        {
          isLoggedIn &&
          <div className='flex filter-section'>
            <Select
              placeholder='Authors'
              value={pathname.includes("/dashboard") ? { value: username, label: username } : selectedAuthor}
              options={authorOptions}
              onChange={onAuthorChange}
              isClearable={true}
              isDisabled={pathname.includes("/dashboard")}
            ></Select>
            <Select
              placeholder='Tags'
              value={selectedTag}
              options={tagOptions}
              onChange={onTagChange}
              isClearable={true}
            ></Select>
            <button 
              className='button button--search' 
              disabled={disableButton}
            >
              <span className={disableButton ? 'hidden' : undefined}>Search</span>
              <span className={disableButton ? 'spinner spinner--button' : undefined} />
            </button>
          </div>
        }
      </form>
    ) 
  }
}

SearchForm.propTypes = {
  handleFilter: PropTypes.func,
  authors: PropTypes.array,
  tags: PropTypes.array,
  history: PropTypes.object,
  handleGetAuthors: PropTypes.func,
  handleGetTags: PropTypes.func,
  className: PropTypes.string,
  isLoggedIn: PropTypes.bool,
  location: PropTypes.object,
  profile: PropTypes.object
};

const mapStateToProps = ({ authors, tags, profile }) => ({ authors, tags, profile });

const mapDispatchToProps = (dispatch) => ({
  handleFilter: (search, tag, author) => dispatch(getFilteredArticles(search, tag, author)),
  handleGetAuthors: () => dispatch(getAuthors()),
  handleGetTags: () => dispatch(getTags())
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchForm);
