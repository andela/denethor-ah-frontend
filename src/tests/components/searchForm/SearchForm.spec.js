import React from 'react';
import { shallow } from 'enzyme';
import Select from 'react-select';

import { SearchForm } from '../../../components/searchForm/SearchForm';
import axios from '../../../utils/axiosConfig';

jest.mock('../../../utils/axiosConfig');

test('Should render search form', () => {
  const props = { location: { pathname: '/' }, profile: { username: undefined } };
  const wrapper = shallow(<SearchForm {...props} />);

  expect(wrapper).toMatchSnapshot();
});

it('Should set searchString state on input change', () => {
  const value = 'foobar';
  const props = { 
    location: { pathname: '/' }, profile: { username: undefined }, isLoggedIn: true 
  };
  const wrapper = shallow(<SearchForm {...props} />);
  

  wrapper.find('input').at(0).simulate('change', { target: { value } });
  expect(wrapper.state('searchString')).toBe(value);
});

it('Should set author state on input change', () => {
  const value = 'foobar';
  const props = { 
    location: { pathname: '/' }, profile: { username: undefined }, isLoggedIn: true 
  };
  const wrapper = shallow(<SearchForm {...props} isSearchOnly={false} />);
  wrapper.find(Select).at(0).props().onChange({ value });
  expect(wrapper.state('selectedAuthor')).toEqual({ value });
});

it('Should set tag state on input change', () => {
  const value = 'foobar';
  const props = { 
    location: { pathname: '/' }, profile: { username: undefined }, isLoggedIn: true 
  };
  const wrapper = shallow(<SearchForm {...props} isSearchOnly={false} />);
  wrapper.find(Select).at(1).props().onChange({ value });
  expect(wrapper.state('selectedTag')).toEqual({ value });
});

it('Should call onSubmit for invalid form submission', () => {
  const props = { 
    location: { pathname: '/' }, profile: { username: undefined }, isLoggedIn: false 
  };
  const wrapper = shallow(<SearchForm {...props} />);
  const preventDefault = jest.fn();

  wrapper.find('form').simulate('submit', { preventDefault });
  expect(preventDefault).toHaveBeenCalled();
});

it('Should call onSubmit for valid form submission', () => {
  const value = 'Search string';
  axios.post.mockImplementationOnce();
  const push = jest.fn();
  const props = { 
    history: { push }, location: { pathname: '/' }, profile: { username: undefined }, isLoggedIn: false 
  }
  const preventDefault = jest.fn();
  const wrapper = shallow(<SearchForm {...props} />);

  wrapper.find('input').at(0).simulate('change', { target: { value } });
  expect(wrapper.state('searchString')).toBe(value);
  wrapper.find('form').simulate('submit', { preventDefault });
  expect(preventDefault).toHaveBeenCalled();
});