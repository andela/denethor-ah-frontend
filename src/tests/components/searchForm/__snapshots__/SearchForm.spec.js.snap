// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render search form 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="search-bar w-100"
  >
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Search for articles..."
      type="text"
      value=""
    />
  </div>
</form>
`;
